src/bootstrap/builder.rs

```diff
@@ -258,8 +259,15 @@ impl StepDescription {
         if paths.is_empty() || builder.config.include_default_paths {
             for (desc, should_run) in v.iter().zip(&should_runs) {
                 if desc.default && should_run.is_really_default() {
-                    for pathset in &should_run.paths {
-                        desc.maybe_run(builder, pathset);
+                    match &builder.config.cmd {
+                        Subcommand::Setup { .. } => {
+                            desc.maybe_run(builder, &PathSet::empty());
```

Comment:
Why does this need to be special-cased?

Reply:
Yes, I agree this is not ideal at all, but I found that if no path is provided when running setup command, i.e. `x setup`, `StepDescrption`'s `maybe_run` will be called 5 times for each of the default profile. 

It seems the working way for me to make it run exactly once, no more no less, is to set this special-case here, to call `maybe_run` once, and not do the `for ... in` loop, for `setup`.

I actually spent quite a while thinking about how to achieve above without special-casing, please let me know if there is an alternative without special-casing!

Reply:
@bentongxyz can you please run `r? @Mark-Simulacrum` so mark gets assigned as the reviewer? I don't have permissions to do it myself.

Reply:
I think the proper fix is to wait until https://github.com/rust-lang/rust/pull/96501 lands so we can pass 0 paths to `make_run`, then do that in the default block for all steps, not just `Setup`.

Reply:
r? @Mark-Simulacrum

Reply:
I think it only works as a top-level comment :( or in the PR description.

Reply:
It looks like the documentation on Step::should_run and Step::make_run is outdated, so that's probably a good idea to update as well (could be in this PR or a separate one).

I think the intent behind calling maybe_run for each of the pathsets configured by a step in should_run is that (historically) test::Crate and similar steps had differing behavior depending on the pathset they were called with, and would run through all the crates one by one with this logic. If we had called with an empty pathset, then presumably that would break those Steps. I think e.g. https://github.com/rust-lang/rust/pull/96501 might affect this though, I'll need to sit down for some time and think through the interactions here I suppose and read the code a little more.

> [...] seems to assume that steps shouldn't care which path they're called with; but I don't see why that should be true?

I am not sure why you think that commit assumes this - indeed, the current logic I think rather goes to show the *opposite* which is that we must call with every path if we're invoking by default, rather than assuming all the paths are equivalent. (Well, not *every* path. Every path *set*, so aliased paths are only invoked with once).